daemon/cmd/settings.go

```diff
@@ -0,0 +1,142 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package cmd
+
+import (
+	"context"
+	"crypto/md5"
+	"fmt"
+	"os"
+	"slices"
+	"strings"
+
+	"github.com/spf13/cast"
+
+	k8sConsts "github.com/cilium/cilium/pkg/k8s/constants"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/option/resolver"
+)
+
+// sourceValue tracks value source origin.
+type sourceValue struct {
+	value, source string
+}
+
+// String returns a source value for humans.
+func (s sourceValue) String() string {
+	return fmt.Sprintf("%s[%s]", s.value, s.source)
+}
+
+// sourceSettings tracks config settings for a given source.
+type sourceSettings map[string]sourceValue
+
+func newSourceSettings(s string, mm map[string]string) sourceSettings {
+	ss := make(map[string]sourceValue, len(mm))
+	for k, v := range mm {
+		ss[k] = sourceValue{value: v, source: s}
+	}
+
+	return ss
+}
+
+func (mm sourceSettings) overlay(desired sourceSettings) sourceSettings {
+	for k, v := range desired {
+		if _, ok := mm[k]; ok {
+			mm[k] = v
+		}
+	}
+
+	return mm
+}
+
+// computeDeltas compute deltas between desired config settings and agent settings.
+func (mm sourceSettings) computeDeltas(settings map[string]string) []string {
+	var dd []string
+	for k, v := range mm {
+		if sv, ok := settings[k]; ok {
+			if v.value != sv {
+				dd = append(dd, fmt.Sprintf("Mismatch for key [%s::%s]: expecting %q but got %q", v.source, k, v.value, sv))
+			}
+		} else {
+			dd = append(dd, fmt.Sprintf("No entry found for key: [%s::%s]", v.source, k))
+		}
+	}
+	slices.Sort(dd)
+
+	return dd
+}
+
+// computeSettingsCheckSum computes config settings checksum.
+func (mm sourceSettings) computeCheckSum() string {
+	kk := make([]string, 0, len(mm))
+	for k := range mm {
+		kk = append(kk, k)
+	}
```

Comment:
nit (feel free to ignore): You could use https://pkg.go.dev/golang.org/x/exp/maps#Keys here if you wanted

Reply:
Brilliant. Thank you for the heads up!